### ISSUE TITLE
Deployments with shield import job get unexpected no-op bosh updates triggering process restarts

### ISSUE THREAD

**Issue by gberche-orange** [orange-cloudfoundry/paas-templates#683]
### Expected behavior

As a paas-templates operator, in order to avoid downtime with deployment having spofs, I expect `bosh deploy` to not trigger job restarts when no changes are detected

### Observed behavior

A bosh deploy without any changes triggers systematically a job update
```
Task 3216352 | 16:47:53 | Preparing deployment: Preparing deployment (00:00:01)
Task 3216352 | 16:47:54 | Preparing deployment: Rendering templates (00:00:03)
Task 3216352 | 16:47:57 | Preparing package compilation: Finding packages to compile (00:00:00)
Task 3216352 | 16:47:57 | Updating instance openldap: openldap/b8136f59-551b-4310-9e96-de1797c89890 (0) (00:00:58)
```

As a result, deployments with SPOFs have downtime when `bosh deploy` is requested such as following a git reset

### Detailed symptoms

The job are actually observed to restart on the VM

```
openldap/b8136f59-551b-4310-9e96-de1797c89890:/var/vcap/bosh/log# less \@400000005e37ec4837a454d4.s

2020-02-03_09:47:21.35812 [Action Dispatcher] 2020/02/03 09:47:21 INFO - Received request with action prepare
2020-02-03_09:47:21.35826 [Action Dispatcher] 2020/02/03 09:47:21 DEBUG - Payload
2020-02-03_09:47:21.35831 ********************
2020-02-03_09:47:21.35838 {"protocol":3,"method":"prepare","arguments":[{"deployment":"openldap","job":{"name":"openldap","templates":[
{"name":"ldap-server","version":"2512a8175c1638a4cef44acc2fa01561e1e5f264","sha1":"39b35adbb9dc34afe42a885c72dacb32eab31db1","blobstore
_id":"564e1f58-62e9-44ec-b2c1-537f1e3e0ff9","logs":[]},{"name":"docker","version":"7e792355e6e9650315b0d2f3547b859c7ea653b859f7d7f52484
f39c27a2afb9","sha1":"sha256:7940205d1ad975c69fdb127183658ed87802760a085444c74e61355db13a2713","blobstore_id":"70dd9348-f424-4d82-b906-
```

The bosh director traces show that configuration changes are systematically detected

```
$ bosh task 3215707 | grep -A1 configuration_changed
D, [2020-02-03T15:03:49.638156 #21623] [] DEBUG -- DirectorJobRunner: configuration_changed? changed FROM: 7ea072f1ec7fc59abf515d93aa4f57025ae7b13b TO: ee6df0cc6dfb248512fe8e190cc6080a2d2bc4c7 on instance openldap/b8136f59-551b-4310-9e96-de1797c89890 (0)

D, [2020-01-29T10:52:44.183774 #13911] [] DEBUG -- DirectorJobRunner: configuration_changed? changed FROM: 046b6cb766d2b3a870a2318cc75680eb5ee39330 TO: 08c4fd2a081faa8e995a775aadbe0fe0bdc70774 on instance openldap/b8136f59-551b-4310-9e96-de1797c89890 (0)
D, [2020-01-29T10:52:44.194126 #13911] [] DEBUG -- DirectorJobRunner: Need to update instance 'openldap/b8136f59-551b-4310-9e96-de1797c89890 (0)', changes: "configuration"
```

This was diagnosed by adding extra debugging support in bosh director

```
$ vi /var/vcap/packages/director/gem_home/ruby/2.6.0/gems/bosh-director-core-0.0.0/lib/bosh/director/core/templates/rendered_job_instance.rb

$ diff /var/vcap/packages/director/gem_home/ruby/2.6.0/gems/bosh-director-core-0.0.0/lib/bosh/director/core/templates/rendered_job_instance.rb  /var/vcap/packages/director/gem_home/ruby/2.6.0/gems/bosh-director-core-0.0.0/lib/bosh/director/core/templates/rendered_job_instance.rb.orig 
17c17
< traceContents=[]
---
> 
29d28
< traceContents<<bound_templates
31,33d29
< traceFile = File.open('/var/vcap/data/director/tmp/digest-trace.txt-'+instance_digest.hexdigest.to_s,"w")
< traceFile.write(traceContents.join)
< traceFile.close

$  monit restart director
$ watch monit summary

$ cd /var/vcap/data/director/tmp

bosh/1c2da1a2-bc97-4f5e-b83b-7db71c2104d1:/var/vcap/data/director/tmp# diff digest-trace.txt-1f3e5da7b0b9c829f9cc10f20eb81bcae55dba2b digest-trace.txt-95b41f1b4285bd7f9990e666e09b09f800b9894a
1487c1487
< $bin/shield revoke-auth-token openldap-openldap-0-1580748398
---
> $bin/shield revoke-auth-token openldap-openldap-0-1580748477
1501c1501
< token=$($bin/shield create-auth-token openldap-openldap-0-1580748398)
---
> token=$($bin/shield create-auth-token openldap-openldap-0-1580748477)

```

Root cause is https://github.com/shieldproject/shield-boshrelease/blob/master/jobs/import/templates/bin/run#L6-L14 which has non deterministic job templates

```erb
<% token_name = p('token_name').gsub('(deployment)', spec.deployment)
                   .gsub('(name)',       spec.name || '')
                   .gsub('(index)',      spec.index.to_s) 
                   .gsub('(time)',       Time.now.to_i.to_s) %>

cleanup() {
<% if p('generate_token') %>
echo "revoke token"
$bin/shield revoke-auth-token <%= token_name %>
exit 0
<% end %>
}
```

### Affected release

Reproduced on version 45.0.0

**Comment by gberche-orange** [comment]
@JCL38-ORANGE Using `(deployment)-(name)-(index)` as token name makes the template rendering deterministic, but not unique across root deployments. Do we have cases where bosh deployment names collides across root deployments ? Coab bosh deployment names should have unique ids

from https://bosh.io/docs/jobs/#properties-spec
> spec.deployment: Name of the BOSH deployment containing this instance.
> spec.name: Name of the instance.

The workaround of setting token_name to backtick hostname backtick, i.e. to inject an os command through a manifest property , such as `hostname`, which resolves to bosh-agent-id may be broken by future security hardening ?

**Comment by poblin-orange** [comment]
@gberche-orange @JCL38-ORANGE 
bosh is unclear on hostname setting (had issues on cfcr for persistence, hostname change on bosh recreate for instance).
I suggest we should not have conflicts between 2 root deployments (bad example master-depls/shield, versus coab-depls/shield)

**Comment by poblin-orange** [comment]
merged for 45.0.2
